File: arrays/subarray_sum/subarray_sum.go
package main

import "fmt"

func main() {
	testCases := []struct {
		input    []int
		window   int
		expected int
	}{
		{[]int{2, 1, 5, 1, 3, 2}, 3, 9},
		{[]int{2, 3, 4, 1, 5}, 2, 7},
		{[]int{1, 2, 3}, 3, 6},
		{[]int{-1, -2, -3, -4}, 2, -3},
		{[]int{5}, 1, 5},
		{[]int{1, 2, 3, 10}, 2, 13},
	}
	for _, tc := range testCases {
		result := bruteForceSubarraySum(tc.input, tc.window)
		status := "failed"
		if result == tc.expected {
			status = "passed"
		}
		fmt.Printf("bruteForceSubarraySum(%v, %d) = (%d) -> %s\n", tc.input, tc.window, result, status)
	}
	for _, tc := range testCases {
		result := optSubarraySum(tc.input, tc.window)
		status := "failed"
		if result == tc.expected {
			status = "passed"
		}
		fmt.Printf("optSubarraySum(%v, %d) = (%d) -> %s\n", tc.input, tc.window, result, status)
	}
}

func bruteForceSubarraySum(nums []int, k int) int {
	sum := func(numbers []int) int {
		total := 0
		for _, x := range numbers {
			total += x
		}
		return total
	}

	maxSum := sum(nums[:k])
	for i := 0; i < len(nums)-k+1; i++ {
		currentSum := 0
		for j := i; j < i+k; j++ {
			currentSum = currentSum + nums[j]
		}
		maxSum = max(currentSum, maxSum)
	}
	return maxSum
}

func optSubarraySum(nums []int, k int) int {
	sum := func(numbers []int) int {
		total := 0
		for _, x := range numbers {
			total += x
		}
		return total
	}
	maxSum := sum(nums[:k])
	currentSum := maxSum
	for j := k; j < len(nums); j++ {
		currentSum += nums[j]
		currentSum -= nums[j-k]
		maxSum = max(currentSum, maxSum)
	}
	return maxSum
}

File: arrays/subarray_sum/subarray_sum.py
def subarray_sum_bruteforce(nums : list[int], k : int) -> int:
    max_sum = sum(nums[:k])
    for i in range(len(nums) - k + 1):
        current_sum = 0
        for j in range(i, i + k):
            current_sum = current_sum + nums[j]
        max_sum = max(current_sum, max_sum)
    return max_sum

def subarray_sum_opt(nums : list[int], k : int) -> int:
    current_sum = sum(nums[:k])
    max_sum = current_sum
    for i in range(k, len(nums)):
        current_sum += nums[i]
        current_sum -= nums[i - k] 
        max_sum = max(max_sum, current_sum)
    return max_sum     

if __name__ == "__main__":
    cases = [
        ([2, 1, 5, 1, 3, 2], 3, 9),
        ([2, 3, 4, 1, 5],    2, 7),
        ([1, 2, 3],          3, 6),
        ([-1,-2,-3,-4],      2, -3),
        ([5],                1, 5),
    ]
for nums, k, expected in cases:
    result = subarray_sum_bruteforce(nums, k)
    status = "Passed" if result == expected else "Failed"
    print(f"{status} subarray_sum_bruteforce({nums}, k={k}) = {result}")

for nums, k, expected in cases:
    result = subarray_sum_opt(nums, k)
    status = "Passed" if result == expected else "Failed"
    print(f"{status} subarray_sum_opt({nums}, k={k}) = {result}")
    